daft/runners/pyrunner.py

```diff
@@ -466,9 +609,17 @@ def _physical_plan_to_partitions(
                                     next_step.partial_metadatas,
                                 )
 
-                            resource_request = next_step.resource_request
+                            def create_resource_release_callback(
+                                resources: AcquiredResources,
+                            ) -> Callable[[futures.Future], None]:
+                                """We use a higher order function here to capture the value of `resources` during the creation of the callback instead of during its call."""
+
+                                def inner(_):
+                                    self._resources.release(resources)
+
+                                return inner
 
-                            future.add_done_callback(lambda _: self._release_resources(resource_request))
+                            future.add_done_callback(create_resource_release_callback(resources))
```

Comment:
I don't quite understand this as to why we need to have this as an inline function.

Could we not do:

```
def _release_resources_callback(self, resources):
    return lambda _: self._resources.release(resources)

...

    future.add_done_callback(self._release_resources_callback(resources))
```